### PR TITLE
Improve determinism/security of CI environment

### PR DESCRIPTION
**Pull request by indygreg**
I wrote detailed commit messages explaining each of the changes I made. GitHub's PR UI (sadly) doesn't reward me for doing so by not giving me an opportunity to easily surface them in the PR description. Please read the commit messages for rationale behind this PR.

The short version is these commits improve the security of Rust's toolchain build environment. The commits are far from a comprehensive solution to the myriad of software supply chain weaknesses I found in Rust's CI. But they are a good start. I hope that with this PR I enlighten Rust maintainers about the existing non-deterministic, non-reproducible nature of Rust's CI and the security risks intrinsic to these properties. My goal is for Rust to commit to deterministic, reproducible CI for toolchain builds (at least as far as it concerns dependencies) so Rust toolchains aren't susceptible to software supply chain attacks via non-deterministic dependencies.

### CODE REVIEW

**Comment by RalfJung** [comment]
> I wrote detailed commit messages explaining each of the changes I made. GitHub's PR UI (sadly) doesn't reward me for doing so by not giving me an opportunity to easily surface them in the PR description. Please read the commit messages for rationale behind this PR.

For this reason and others, I usually prefer comments in the code. They are also much more likely to be read by future developers editing this code.

Of course, there's not always a great place to put that comment.

File: src/ci/docker/scripts/secure-download.sh [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  exit 1
+fi
+
+URL=$1
+SHA256=$2
+BASENAME=`basename ${URL}`
+
+ON_EXIT="rm -f ${BASENAME}"
+
+trap "rm -f ${BASENAME}" EXIT
```
Comment (RalfJung) [inline src/ci/docker/scripts/secure-download.sh]: Wow, so one needs *both* `ON_EXIT` and `trap` to reliably clean up? Shell scripting makes me sad. :(

Might be worth adding a comment with (a link to) some explanation why, if a good one exists.

**Comment by pietroalbini** [comment]
Thanks for sending this PR, it's deeply appreciated!

We're aware that this issue with our CI exists and we started fixing it a long time ago on a case-by-case basis when things we downloaded from CI broke or went missing (like 404s or the Android SDK automatically updating itself). Due to higher priority stuff coming up all the time though we never did a full audit of all the things we download, and it's great that you're interested in this!

The approach we'd prefer is a bit different than what you're suggesting: your solution is great to prevent malicious changes to the things we download, but it doesn't do anything to prevent the links from breaking, which is something that unfortunately happens quite often.

What we've been doing so far is vendoring all the binaries in an [S3 bucket](https://rust-lang-ci2.s3.amazonaws.com/rust-ci-mirror) and fetching them from it instead of the upstream source: this solves both of the problems we have, as we can guarantee the links will be up in the future.

Would you be willing to update all the links to point to that S3 bucket? Of course you can't upload stuff in the bucket yourself, but if you go through the config and change everything to point to the bucket I can then upload things to it. For example:

```bash
# Source URL: https://www.openssl.org/source/old/1.0.2/openssl-1.0.2n.tar.gz
curl https://rust-lang-ci2.s3.amazonaws.com/rust-ci-mirror/openssl-1.0.2n.tar.gz
```

> The commits are far from a comprehensive solution to the myriad of software supply chain weaknesses I found in Rust's CI.

If you found other weaknesses other than this one please follow [our security policy](https://www.rust-lang.org/policies/security), thanks!

**Comment by Dylan-DPC** [comment]
ping from triage @indygreg any updates on this?

**Comment by joelpalmer** [comment]
Hi @indygreg - this is a ping from Triage. Closing this PR due to prolonged inactivity. Please re-open before pushing changes. Thank you for the PR. 

@rustbot modify labels to +S-inactive-closed and -S-waiting-on-author

**Pull request closed without merging**